### DISCUSSION TITLE
Migration of the source code to Codeberg

### DISCUSSION THREAD

**Discussion by linux-lukas** [owncast/owncast#1957] [General Owncast discussion]
Given the dependencies on Microsoft in 2018 after it bought GitHub, it makes me wonder if the source code might not also be on Codeberg, a Gitea-based GitHub and GitLab alternative?

**Comment by linux-lukas** [comment]
With more detail: 

### A. Problem / Goal
Since the purchase of GitHub by Microsoft in 2018, a dependence on the BigTech corporation can no longer be denied.

On the one hand, I can understand why GitHub was chosen as the platform for making source code available: "Everyone is here".

On the other hand, I see the danger of a vendor-lockin effect and that open source projects become centrally dependent on Microsoft. In my eyes, this is very dangerous for free and open source software and hardware projects.

In the medium and long term, the goal would be to become independent of GitHub and thus of Microsoft. The Gitea-based Codeberg project of Codeberg e. V. in Berlin would be a good choice here.

There are also (legal) problems with compliance with the licence of GitHub functions, such as the co-pilot.

### B. Solution
My considered solution to the problem described in A. would be the following:

A user of this open source project creates a user account on https://codeberg.org/
If necessary: This user creates an organisation for the project.
A "personal access token" is created on the GitHub account, which has appropriate rights to the organisation repositories, using the developer options in the settings.
all repositories would be migrated with this access token into the ownership of the organisation created in step two.
Regarding step four, there is an entry in the documentation of Codeberg: https://docs.codeberg.org/advanced/migrating-repos/

### C. Alternatives
A possible alternative would be to perform the first three steps as described in B. A possible alternative would be to perform the first three steps as described in B., and modify the fourth step to include a mirror of GitHub. So that all issues and such that would be created in the GitHub repository would be transferred to the Codeberg repository.

### D. Responsibilities
I would see the responsibility in the owners of the repository and, if necessary, additional project participants. 

cc: @gabek @jeyemwey @muesli

### E. Other
Basically, a look at the documentation of Codeberg is not unwise: https://docs.codeberg.org/

Should it be necessary to manage repositories in organisations, this is also possible under Codeberg, see: https://docs.codeberg.org/collaborating/create-organization/

Regarding licensing there is a page in the documentation of Codeberg: https://docs.codeberg.org/getting-started/licensing/

Funfact 1: The Pixelfed maintainer has written on Mastodon that he is considering migrating the Pixelfed project away from GitHub, to Codeberg: https://mastodon.social/@dansup/10869007146309141

Funfact 2: Aral had migrated - also on Mastodon - to Codeberg for his GitHub projects as well as those of his Foundation, which he organised together with another person, Small Technology, or is still migrating: https://mastodon.ar.al/@aral/108651118032076133

**Comment by gabek** [comment]
While I appreciate your thoughts, Codeberg is also a centralized source code repository provider. I trust them as much as I trust GitHub, that being not very much at all. I don't know them. I don't know how they'll screw me over tomorrow, just like I don't know how eventually GitHub will do something to push the project away.

And when that day comes I'll have to spin up our own Gitea instance for managing the project. You can't trust anybody but yourself.

Given that I don't want to start self-hosting these services right now, and moving to yet another huge, centralized service provider is the alternative, Owncast using GitHub will have to be good enough. It's very much a "devil you know" scenario. For now Microsoft lets me get work done and work with contributors.  Eventually they'll do something truly horrible and GitHub won't be a viable option anymore, and then I guess I'll be running my own project management and source control software somewhere.

**Reply by linux-lukas** [reply]
Well, I can understand some of the concerns about Codeberg, although the Codeberg project is based on Gitea and the federation as it is known from Fediverse already works in parts - what I want to say is that in the Gitea project there are efforts to remove the centralisation and to decentralise using Fediverse as an example. 

In addition, Codeberg is run by a non-profit association based in the German capital Berlin, and acting more ethically than GitHub is firmly rooted in the association's DNA. If there are concerns, it is possible to bring light into the darkness in a constructive way.

Nevertheless, I can also understand their points to some extent. GitHub is currently at least still an incredible platform that many know and use and offers many functions and Gitea is still lagging behind in parts, but in the medium and long term it would still not be unwise IMO to consider a change away from GitHub.

Anyway: If this should work out with the federation, a cooperation of an own Gitea instance and e.g. Codeberg would also be possible, which would dissolve many barriers in the area of software development.

**Comment by gabek** [comment]
I wanted to link these two threads: https://github.com/owncast/owncast/issues/3176
